pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go

```diff
@@ -218,102 +209,64 @@ func (ca *cloudCIDRAllocator) Run(stopCh <-chan struct{}) {
 	}
 
 	for i := 0; i < cidrUpdateWorkers; i++ {
-		go ca.worker(stopCh)
+		go wait.UntilWithContext(ctx, ca.runWorker, time.Second)
 	}
 
 	<-stopCh
 }
 
-func (ca *cloudCIDRAllocator) worker(stopChan <-chan struct{}) {
-	for {
-		select {
-		case workItem, ok := <-ca.nodeUpdateChannel:
-			if !ok {
-				klog.Warning("Channel nodeCIDRUpdateChannel was unexpectedly closed")
-				return
-			}
-			if err := ca.updateCIDRAllocation(workItem); err == nil {
-				klog.V(3).Infof("Updated CIDR for %q", workItem)
-			} else {
-				klog.Errorf("Error updating CIDR for %q: %v", workItem, err)
-				if canRetry, timeout := ca.retryParams(workItem); canRetry {
-					klog.V(2).Infof("Retrying update for %q after %v", workItem, timeout)
-					time.AfterFunc(timeout, func() {
-						// Requeue the failed node for update again.
-						ca.nodeUpdateChannel <- workItem
-					})
-					continue
-				}
-				klog.Errorf("Exceeded retry count for %q, dropping from queue", workItem)
-			}
-			ca.removeNodeFromProcessing(workItem)
-		case <-stopChan:
-			return
-		}
-	}
+func (ca *cloudCIDRAllocator) AllocateOrOccupyCIDR(node *v1.Node) error {
+	klog.V(4).Infof("Putting node %s into the work queue", node.Name)
+	ca.queue.Add(node.Name)
+	return nil
 }
 
-func (ca *cloudCIDRAllocator) insertNodeToProcessing(nodeName string) bool {
-	ca.lock.Lock()
-	defer ca.lock.Unlock()
-	if _, found := ca.nodesInProcessing[nodeName]; found {
-		return false
+func (ca *cloudCIDRAllocator) runWorker(ctx context.Context) {
+	for ca.processNextItem(ctx) {
 	}
-	ca.nodesInProcessing[nodeName] = &nodeProcessingInfo{}
-	return true
 }
 
-func (ca *cloudCIDRAllocator) retryParams(nodeName string) (bool, time.Duration) {
-	ca.lock.Lock()
-	defer ca.lock.Unlock()
-
-	entry, ok := ca.nodesInProcessing[nodeName]
-	if !ok {
-		klog.Errorf("Cannot get retryParams for %q as entry does not exist", nodeName)
-		return false, 0
+func (ca *cloudCIDRAllocator) processNextItem(ctx context.Context) bool {
+	key, quit := ca.queue.Get()
+	if quit {
+		return false
 	}
 
-	count := entry.retries + 1
-	if count > updateMaxRetries {
-		return false, 0
-	}
-	ca.nodesInProcessing[nodeName].retries = count
+	defer ca.queue.Done(key)
```

Comment:
When you do this, does it also affect the key if it is requeued when you do ca.queue.AddRateLimited(key)?

What if the key is 'forgotten'? I am worried about race conditions or missed processing.

Reply:
The workqueu guarantees keys are processed only once in each worker

Reply:
Then won't this mark a key that has been requeued as 'done'?

Reply:
best thing is to follow the example https://github.com/kubernetes/client-go/blob/master/examples/workqueue/main.go , I never remember the details 😄